### PR TITLE
Dynamic concept relationships

### PR DESCRIPTION
**Pull request by MiroslavGatsanoga**
# Description

## What

Define dynamic concept relationships in the ontology package.

## Why

[JIRA](https://financialtimes.atlassian.net/browse/UPPSF-2469)

## Anything, in particular, you'd like to highlight to reviewers

Here is how would an example relationships config would look like:
```
relationships:
  HAS_BROADER:
    conceptField: "broaderUUIDs"
  HAS_ORGANISATION:
    conceptField: "organisationUUID"
    oneToOne: true
    neoCreate: true
  HAS_INDUSTRY_CLASSIFICATION:
    conceptField: "naicsIndustryClassifications"
    neoCreate: true
    toNodeWithLabel: "NAICSIndustryClassification"
    properties: ["rank"]
  HAS_ROLE:
    conceptField: "membershipRoles"
    neoCreate: true
    properties:
      [
        "inceptionDate",
        "terminationDate",
        "inceptionDateEpoch",
        "terminationDateEpoch",
      ]
```

## Scope and particulars of this PR (Please tick all that apply)

- [X] Tech hygiene (dependency updating & other tech debt)
- [ ] Bug fix
- [X] Feature
- [ ] Documentation
- [ ] Breaking change
- [ ] Minor change (e.g. fixing a typo, adding config)

___
This Pull Request follows the rules described in our [Pull Requests Guide](https://github.com/Financial-Times/upp-docs/tree/master/guides/pr-guide)

### CODE REVIEW

File: concepts/concepts_service.go [hunk 1]
```diff
@@ -248,6 +183,207 @@ func (s *ConceptService) read(uuid string, transID string) (ontology.NewAggregat
 	return newAggregatedConcept, true, nil
 }
 
+func getReadStatement() string {
```
Comment (ivanruski) [inline concepts/concepts_service.go:186]: **thought:** Maybe we can add a test which calls neo4j with `EXPLAIN`. This will tells us that at least the query is valid.

File: concepts/concepts_service.go [hunk 2]
```diff
@@ -804,152 +959,68 @@ func createNodeQueries(concept ontology.NewConcept, uuid string) []*neoism.Cyphe
...
-				"fiCode":  concept.FigiCode,
-				"orgUUID": concept.IssuedBy,
-			},
-		}
-		queryBatch = append(queryBatch, writeFinIns)
-	}
+	relIDs = filterSlice([]string{concept.PersonUUID})
+	queryBatch = append(queryBatch, createRelQueries(concept.UUID, relIDs, "HAS_MEMBER", true)...)
 
-	if uuid != "" && concept.ParentOrganisation != "" {
```
Comment (ivanruski) [inline concepts/concepts_service.go:861]: **question:** Currently we are not validating whether the source concepts in the request body have `uuid` (I've tried to send a request where the source concept didn't have uuid and I got 200). Should we have a check for this ?

File: concepts/concepts_service.go [hunk 3]
```diff
@@ -960,31 +1031,65 @@ func createNodeQueries(concept ontology.NewConcept, uuid string) []*neoism.Cyphe
...
 	}
+
 	queryBatch = append(queryBatch, createConceptQuery)
 	return queryBatch
 }
 
-//Add relationships to concepts
-func addRelationship(conceptID string, relationshipIDs []string, relationshipType string, queryBatch []*neoism.CypherQuery) []*neoism.CypherQuery {
+// createRelQueries creates relationships Cypher queries for concepts
+func createRelQueries(conceptID string, relationshipIDs []string, relationshipType string, shouldCreate bool) []*neoism.CypherQuery {
```
Comment (ivanruski) [inline concepts/concepts_service.go:1044]: **praise:** This function removed a lot of boilerplate code.

File: concepts/neo_concept.go [hunk 4]
```diff
@@ -241,6 +285,29 @@ func filterSlice(a []string) []string {
 	return a
```
Comment (ivanruski) [inline concepts/neo_concept.go:285]: **issue:** I might be wrong but I think that at the end of the method we have to `return r` instead of `a`.

Comment (ivanruski) [inline concepts/neo_concept.go:285]: **issue:** I might be wrong but I think that at the end of the method we have to `return r` instead of `a`.

File: ontology/config.go [hunk 5]
```diff
@@ -1,26 +1,31 @@
...
+	ConceptField    string   `yaml:"conceptField"`
+	OneToOne        bool     `yaml:"oneToOne"`
+	NeoCreate       bool     `yaml:"neoCreate"`
+	Properties      []string `yaml:"properties"`
+	ToNodeWithLabel string   `yaml:"toNodeWithLabel"`
+}
+
 type Config struct {
-	FieldToNeoProps map[string]string `yaml:"fieldToNeoProps"`
+	FieldToNeoProps map[string]string             `yaml:"fieldToNeoProps"`
```
Comment (ivanruski) [inline ontology/config.go:18]: **question:** Unrelated to this PR but should we have a notion of required and optional props ?

File: concepts/concepts_service.go (hunk 1 above)
Comment (epavlova) [reply concepts/concepts_service.go:186]: **thought:** On the same note I was thinking that in this function and in the functions it calls there are a lot of string manipulations that are very tricky and some corner cases can easily be missed. I was thinking on adding some unit tests for verifying the result of `getReadStatement()` with mocked ontology. Or different kind of testing.

**Review by epavlova — COMMENTED** [review]
I have a few thoughts only.

File: concepts/concepts_service.go (hunk 1 above)
Comment (MiroslavGatsanoga) [reply concepts/concepts_service.go:186]: @ivanruski I am not too keen on using `EXPLAIN` because it may lead to brittle test e.g. every schema change would mean the test will fail. 
@epavlova I will add a test for `getReadStatement()` that shows that we get the same query as in the `Make the Cypher read query more organized` commit.

File: concepts/concepts_service.go (hunk 2 above)
Comment (MiroslavGatsanoga) [reply concepts/concepts_service.go:861]: By "currently" do you mean that it is not validating even before this PR or did I introduced lack of validation in this PR?

File: concepts/neo_concept.go (hunk 4 above)
Comment (MiroslavGatsanoga) [reply concepts/neo_concept.go:285]: Oh, good catch, but given that this function was introduced 3 years ago and we noticed the issue now I am wondering if I should fix it 😆. 
Will fix it and see what happens .

File: ontology/config.go (hunk 5 above)
Comment (MiroslavGatsanoga) [reply ontology/config.go:18]: What do you mean by that, can you give an example?

File: ontology/config.go [hunk 6]
```diff
@@ -1,26 +1,31 @@
...
+	"embed"
 
 	"gopkg.in/yaml.v2"
 )
 
+type RelationshipConfig struct {
+	ConceptField    string   `yaml:"conceptField"`
+	OneToOne        bool     `yaml:"oneToOne"`
+	NeoCreate       bool     `yaml:"neoCreate"`
+	Properties      []string `yaml:"properties"`
```
[the remark this answers is not in the document]
Comment (MiroslavGatsanoga) [reply ontology/config.go:13]: There was a [similar comment](https://github.com/Financial-Times/concepts-rw-neo4j/pull/73#discussion_r658531673) regarding the types of the node properties' values, this additional validation will be done when we actually transition to the ontology.

File: concepts/neo_concept.go (hunk 4 above)
Comment (ivanruski) [reply concepts/neo_concept.go:285]: 😄 .. I though that this function was written recently

File: ontology/config.go (hunk 5 above)
Comment (ivanruski) [reply ontology/config.go:18]: For example:
```
fieldToNeeProps:
  prefUUID:
    required: true
```
And if it's missing we won't ingest the concept.

File: concepts/concepts_service.go (hunk 1 above)
Comment (MiroslavGatsanoga) [reply concepts/concepts_service.go]: I have added a getDeleteStatement() function, extracted both functions that generate Cypher statements in a separate file and added tests leveraging golden files so that we can see how do the generated statements look like.

File: concepts/neo_concept.go [hunk 7]
```diff
@@ -198,14 +197,59 @@ func (nc neoConcept) ТоOntologyNewConcept() (ontology.NewConcept, error) {
...
+
 	return ontology.NewConcept{
+		Relationships:                filterRelationships(rels),
 		Authority:                    nc.Authority,
 		AuthorityValue:               nc.AuthorityValue,
 		BroaderUUIDs:                 filterSlice(nc.BroaderUUIDs),
 		SupersededByUUIDs:            filterSlice(nc.SupersededByUUIDs),
 		FigiCode:                     nc.FigiCode,
 		IssuedBy:                     nc.IssuedBy,
-		LastModifiedEpoch:            nc.LastModifiedEpoch,
```
Comment (ivan-p-nikolov) [inline concepts/neo_concept.go:208]: **question:** We don't copy the timestamp because it is always set in the `setProps`. Is that correct?

File: concepts/neo_concept.go (hunk 4 above)
Comment (ivan-p-nikolov) [reply concepts/neo_concept.go:285]: If I remember correctly, originally `filterSlice` was used when we read from Neo4j. So at most this fix will break some tests.

File: ontology/config.go (hunk 5 above)
Comment (MiroslavGatsanoga) [reply ontology/config.go:18]: We've discussed this but to have it here as well: any required fields or relationships are just expressed through code and not through the config, we could add some config that can be used by the `generic-concept-transformer` to check and verify some fields but that should not be done in this PR.

File: concepts/neo_concept.go (hunk 7 above)
Comment (MiroslavGatsanoga) [reply concepts/neo_concept.go:208]: It was done in [this commit](https://github.com/Financial-Times/concepts-rw-neo4j/pull/74/commits/765887625f373f0f11052439a3e8dbb37f55306f), the idea was to extract the "cleaning" from the `cleanNewConcept` function and making it become `sortSources`.

File: concepts/concepts_service.go [hunk 8]
```diff
@@ -25,6 +25,65 @@ const (
 
 var concordancesSources = []string{"ManagedLocation", "Smartlogic"}
 
+var relationships = map[string]ontology.RelationshipConfig{
```
Comment (ivan-p-nikolov) [inline concepts/concepts_service.go:28]: **suggestion(non-blocking):** I would expect this map to be in `cypher_statements.go`. As I don't see it used in `concetps_services.go` ‏

**Review by ivan-p-nikolov — APPROVED** [review]
LGTM

**Review by ivanruski — APPROVED** [review]

File: concepts/concepts_service.go (hunk 8 above)
Comment (MiroslavGatsanoga) [reply concepts/concepts_service.go:28]: Yes, I agree but in the end after we migrate all relationships this variable will not be needed so it will not matter.
